R/permutation.R

```diff
@@ -0,0 +1,703 @@
+    rstattract <- rstatcalc %>%
+        filter(trt==1) %>%
+        dplyr::select( -!!as.name(tx_col), -trt )
+
+    # Make table of main results with permutation p-values and add
+    # indicator of which years are treated and which not.
+    main_results <- rstattract %>%
+        full_join( pvalues, by=time_col ) %>%
+        mutate( tx = ifelse( !!as.name(time_col) >= treat_year, 1, 0 ) )
+
+
+    # Clean up column ordering and make look nice
+    main_results = main_results %>%
+        relocate( !!as.name(time_col), tx, Yobs ) %>%
+        dplyr::select( -RMSPE )
+
+    # Bundle and return all results
+    list( MDES_table = main_results,
+          RMSPEs = RMSPEs,
+          rstatcalc=rstatcalc )
+}
+
+
+#' Calculate RMSPE for all units on lagged outcomes
+#'
+#' This calculates the RMSPE of the difference between estimated and
+#' observed outcome, averaged across the pre-treatment years, for each
+#' county
+#'
+#' @param lest Entity by year estimate estimates for all treatment and control entities
+#'   (original tx and all the comparisons as pseudo-treated).  Columns
+#'   of entity ID (tx_col), treatment flag (trt), time period (time_col),
+#'   and estimated tx-synthetic control difference (ATT).
+#' @param treat_year The time period in which the treatment was introduced.
+#' @param  tx_col The name to the column containing the treatment variable.
+#' @param time_col The name of the column containing the time variable.
+#'
+#' @noRd
+calculate_RMSPE = function( lest, treat_year, tx_col, time_col ) {
+    stopifnot( !is.null( lest$ATT ) )
+
+    RMSPEs = lest %>% filter( !!as.name(time_col) < treat_year ) %>%
+        group_by( !!as.name(tx_col) ) %>%
+        summarise( RMSPE = sqrt( mean( ATT^2 ) ), .groups = "drop" )
+
+    return(RMSPEs)
+}
+
+#' Estimate robust SE
+#'
+#' Given a list of impact estimates (all assumed to be estimating a
+#' target impact of 0), calculate the modified standard deviation by
+#' looking an in inter-quartile range rather than the simple standard
+#' deviation.  This reduces the effect of small numbers of extreme
+#' outliers, and focuses on central variation.  This then gets
+#' converted to a standard error (standard deviation of the results,
+#' in effect.)
+#'
+#' @param placebo_estimates The list of impact estimates to use for
+#'   estimating the SE.
+#' @param k Number of obs to drop from each tail (so 2k obs dropped
+#'   total)
+#' @param beta Proportion of obs to drop from each tail. E.g., 95\% to
+#'   drop 5\% most extreme.  Only one of beta or k can be non-null.
+#' @param round_beta TRUE means adjust beta to the nearest integer
+#'   number of units to drop (the Howard method).  FALSE means
+#'   interpolate quantiles using R's quantile function.
+#'
+#' @noRd
+estimate_robust_SE = function( placebo_estimates, k = NULL, beta = NULL,
+                               round_beta = FALSE ) {
+
+    stopifnot("Either `k` or `beta` must be NULL." = is.null(k) != is.null( beta ) )
+
```

Comment:
Ditto. This is an internal function, can both be null due to the user putting something in wrong?